### PR TITLE
feat: appdir option (smaller version)

### PR DESCRIPTION
**Pull request by juliusmarminge**
smaller, less opinionated version of #1446

This alternative does not include:
- Server Actions
- ShadCN-ish tailwind config
- Custom UI Button component with twMerge

### CODE REVIEW

File: cli/src/installers/dependencyVersionMap.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export const dependencyVersionMap = {
   "drizzle-orm": "^0.28.5",
   "drizzle-kit": "^0.19.13",
   "dotenv-cli": "^7.3.0",
+  "mysql2": "^3.6.1",
```
Comment (c-ehrlich) [inline cli/src/installers/dependencyVersionMap.ts:19]: required for drizzle-kit as far as i can tell

File: cli/src/installers/drizzle.ts [hunk 2]
```diff
@@ -55,6 +55,7 @@ export const drizzleInstaller: Installer = ({
   packageJsonContent.scripts = {
     ...packageJsonContent.scripts,
     "db:push": "dotenv drizzle-kit push:mysql",
+    "db:studio": "dotenv drizzle-kit studio",
```
Comment (c-ehrlich) [inline cli/src/installers/drizzle.ts:58]: running regular `pnpm drizzle-kit studio` causes the same env issue as we had with push before

File: cli/src/installers/dependencyVersionMap.ts (hunk 1 above)
Comment (juliusmarminge) [reply cli/src/installers/dependencyVersionMap.ts:19]: Yea i think i recall needing that in my devdeps 👍

File: cli/src/installers/drizzle.ts (hunk 2 above)
Comment (juliusmarminge) [reply cli/src/installers/drizzle.ts:58]: 👍👍

Comment (c-ehrlich) [reply cli/src/installers/drizzle.ts:58]: done

File: cli/src/installers/dependencyVersionMap.ts (hunk 1 above)
Comment (AndriiSherman) [reply cli/src/installers/dependencyVersionMap.ts]: We are going to bundle all driver dependencies in drizzle-kit, so this one should become unnecessary

Comment (AndriiSherman) [reply cli/src/installers/dependencyVersionMap.ts]: I'll try to push this next, so you don't need to install anything

File: cli/template/extras/src/app/api/trpc/[trpc]/route.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import { env } from "~/env.mjs";
+import { appRouter } from "~/server/api/root";
+import { createTRPCContext } from "~/server/api/trpc";
+
+const handler = (req: NextRequest) =>
+  fetchRequestHandler({
+    endpoint: "/api/trpc",
+    req,
+    router: appRouter,
+    createContext: () => createTRPCContext({ req }),
```
Comment (ajdiyassin) [inline cli/template/extras/src/app/api/trpc/[trpc]/route.ts:13]: I'm getting this error from `req` passed to the context
`Type 'NextRequest' is missing the following properties from type 'NextApiRequest': query, env, aborted, httpVersion, and 67 more.`
`The expected type comes from property 'req' which is declared here on type 'CreateNextContextOptions'`

**Comment by CombeeMike** [comment]
I'm really interested in how an app started from t3 using app dir looks like & don't want to wait for this PR to be finished 😉
Is there a way I can "manually" create a new project from the latest version in the `appdir-smaller` branch?

I already checked out the repo & tried running `npm run start:cli` or `dev:cli` but both resulted in "No tasks were executed as part of this run.".

Do I need to provide some cmd line arguments or do I need to run a different cmd?

I don't want to use `pnpm create t3-app@7.20.3-beta.61030b4` which was mentioned in [this comment](https://github.com/t3-oss/create-t3-app/pull/1567#issuecomment-1717359142) as there had been some changes since then which I'd like to have included.

**Comment by juliusmarminge** [comment]
> I don't want to use `pnpm create t3-app@7.20.3-beta.61030b4` which was mentioned in [this comment](https://github.com/t3-oss/create-t3-app/pull/1567#issuecomment-1717359142) as there had been some changes since then which I'd like to have included.


That comment is updated each time a release is cut so that is the latest

**Comment by juliusmarminge** [comment]
I'll pull one right now and you should see thst comment getting updated in just a minute or two

**Comment by CombeeMike** [comment]
Is this supposed to build ATM?

It doesn't for me. `npm run build` right after bootstrapping (+ adding some env data), gives me the following:

```
Generating static pages (0/4)  [=   ] << query  #1 post.hello  {
  input: { text: 'from tRPC' },
  result: s [TRPCClientError]: Dynamic server usage: Page couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
      at s.from (/home/mike/dev/test-app/.next/server/chunks/102.js:1:21541)
      at /home/mike/dev/test-app/.next/server/chunks/102.js:1:25381 {
    meta: undefined,
    shape: undefined,
    data: undefined,
    [cause]: n [Error]: Dynamic server usage: Page couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
        at i (/home/mike/dev/test-app/.next/server/chunks/175.js:1:3588)
        at d (/home/mike/dev/test-app/.next/server/chunks/175.js:1:1083)
        at Object.headers (/home/mike/dev/test-app/.next/server/app/page.js:1:3490)
        at Object.headers (/home/mike/dev/test-app/.next/server/chunks/102.js:5:4719)
        at g (/home/mike/dev/test-app/.next/server/chunks/102.js:1:22751)
        at k (/home/mike/dev/test-app/.next/server/chunks/102.js:5:3963)
        at Object.fetch (/home/mike/dev/test-app/.next/server/chunks/102.js:5:4613)
        at Timeout.o (/home/mike/dev/test-app/.next/server/chunks/102.js:1:24229)
        at listOnTimeout (node:internal/timers:569:17)
        at process.processTimers (node:internal/timers:512:7) {
      digest: 'DYNAMIC_SERVER_USAGE'
    }
  },
  elapsedMs: 5
}
 << query  #2 post.hello  {
  input: { text: 'from tRPC' },
  result: s [TRPCClientError]: Dynamic server usage: Page couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
      at s.from (/home/mike/dev/test-app/.next/server/chunks/102.js:1:21541)
      at /home/mike/dev/test-app/.next/server/chunks/102.js:1:25381 {
    meta: undefined,
    shape: undefined,
    data: undefined,
    [cause]: n [Error]: Dynamic server usage: Page couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
        at i (/home/mike/dev/test-app/.next/server/chunks/175.js:1:3588)
        at d (/home/mike/dev/test-app/.next/server/chunks/175.js:1:1083)
        at Object.headers (/home/mike/dev/test-app/.next/server/app/page.js:1:3490)
        at Object.headers (/home/mike/dev/test-app/.next/server/chunks/102.js:5:4719)
        at g (/home/mike/dev/test-app/.next/server/chunks/102.js:1:22751)
        at k (/home/mike/dev/test-app/.next/server/chunks/102.js:5:3963)
        at Object.fetch (/home/mike/dev/test-app/.next/server/chunks/102.js:5:4613)
        at Timeout.o (/home/mike/dev/test-app/.next/server/chunks/102.js:1:24229)
        at listOnTimeout (node:internal/timers:569:17)
        at process.processTimers (node:internal/timers:512:7) {
      digest: 'DYNAMIC_SERVER_USAGE'
    }
  },
  elapsedMs: 1
}
```

Am I doing something wrong here or is this simply not ready to build yet?

**Comment by juliusmarminge** [comment]
CI builds - last time I tried it built

**Comment by CombeeMike** [comment]
Just tried in 3 different envs:
* Ubuntu 22.04.2 LTS in WSL2, npm v9.5.1, Node v18.16.1
* "Real" Ubuntu 22.04.3 LTS, npm v9.6.4, Node v18.17.1
* Windows 10 cmd line, npm v8.1.4, Node v16.17.0

Following steps:
* `npm create t3-app@7.20.3-beta.2e8399f` with options:
  `TS: yes`, `TW: yes`, `tRPC: yes`, `NextAuth`, `Drizzle`, `App dir: yes`, `Git init: no`, `npm i: yes`, `Alias: default ~/`
* Write some dummy values into properties `DATABASE_URL` & `NEXTAUTH_SECRET` in `.env`
* `npm run build`

All 3 resulted in the same error mentioned above 🤷‍♂️

**Comment by LZL0** [comment]
I would love to see this working. Please let me know how I could assist.

Next.js documentation: [https://nextjs.org/docs/messages/app-static-to-dynamic-error](https://nextjs.org/docs/messages/app-static-to-dynamic-error)
From first glance it seems to be having issues with the `headers()` function invocation in `cli/template/extras/src/app/layout/with-trpc.tsx` and `cli/template/extras/src/app/layout/with-trpc-tw.tsx`.

**Comment by CombeeMike** [comment]
Seems to be the line `const heads = new Map(headers());` in `trpc/server.ts`. I added some logs and the one before this invocation to `headers()` shows up in my build log right before the error messages:

**trpc/server.ts**
```ts
    // ...
    unstable_httpBatchStreamLink({
      url: getUrl(),
      headers() {
        console.log('creating headers...'); // <- Shows in log right before error
        const heads = new Map(headers());
        heads.set("x-trpc-source", "rsc");
        console.log('headers created'); // <- Not showing in logs. Guess `headers()` invocation above throws
        return Object.fromEntries(heads);
      },
    }),
    // ...
```

**layout.tsx**
```
// ...
<body className={`${robotoMono.variable} ${roboto.variable} font-sans`}>
  <AppProviders
    headers={(() => {
      console.log('creating headers in layout...'); // <- Not showing in log, maybe we're not getting this far?
      return headers();
    })()}
  >
  // ...
</body>
// ...
```

**Build log output**
```
...
Generating static pages (1/4)  [==  ] 
creating headers...
<< query  #1 post.hello  {
  input: { text: 'from tRPC' },
  result: s [TRPCClientError]: Dynamic server usage: Page couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error
...
```

Whilst this seems rather clear, I don't know how to work around this as I don't have enough knowledge about what this `unstable_httpBatchStreamLink` is actually doing and whether we could get the actual header data from somewhere else, maybe create it in some closure before the async context etc. 🤷‍♂️

**Comment by juliusmarminge** [comment]
> Whilst this seems rather clear, I don't know how to work around this as I don't have enough knowledge about what this unstable_httpBatchStreamLink is actually doing and whether we could get the actual header data from somewhere else, maybe create it in some closure before the async context etc. 

Hmm, I'll try to have a look at this in the coming days, but we're using this setup at work and it's fine. Afaik, this is the only way to read the incoming request headers for client components during ssr :/

**Comment by CombeeMike** [comment]
`export const dynamic = 'force-dynamic';` in my main `page.tsx` seems to be doing its job 🎉
At least it builds now and I get the correct page after `npm run start`.

I'm not sure about the actual downsides and whether or not this should be in the default `page.tsx` created from the template as I'm rather new to the whole app router but from quickly going over the docs, it seems to be the right thing at least for my current use case anyways.

Thanks for the support 🙏, I hope it's ok that I "hijacked" this PR in the way I did 🙈

**Comment by LZL0** [comment]
I believe by adding `force-dynamic` we are forcing SSR (server-side rendering).

**Comment by zihaolam** [comment]
Used pnpm create t3-app@7.20.3-beta.61030b4, did not make any code changes and ran dev server and met this issue with the server side trpc call.

Screenshot 2023-10-11 at 2 03 21 AM

**Comment by c-ehrlich** [comment]
> Used pnpm create t3-app@7.20.3-beta.61030b4, did not make any code changes and ran dev server and met this issue with the server side trpc call.
> 
> Screenshot 2023-10-11 at 2 03 21 AM

thanks for the report. this was due to using a non-3000 port. just pushed a change to make it work on any port.

be aware though that next-auth will still require a specific port for most providers including the default discord one

**Comment by zihaolam** [comment]
> > Used pnpm create t3-app@7.20.3-beta.61030b4, did not make any code changes and ran dev server and met this issue with the server side trpc call.
> > Screenshot 2023-10-11 at 2 03 21 AM
> 
> thanks for the report. this was due to using a non-3000 port. just pushed a change to make it work on any port.
> 
> be aware though that next-auth will still require a specific port for most providers including the default discord one

Thanks for the prompt reply & fix! How do i incorporate the latest changes to the current project? do i need to bootstrap a new one?

**Comment by c-ehrlich** [comment]
easiest is probably just looking at the diff from the last commits

**Comment by czystyl** [comment]
Hi @juliusmarminge, I've tested `pnpm create t3-app@7.20.3-beta.61030b4`, but I can see that behavior inside the Client Component useQuery does not use cache. You can see it when I'm typing; it goes back to the initial data from the RSC or is undefined when using it without. 

Am I missing something? 

https://github.com/t3-oss/create-t3-app/assets/10349378/21673450-8313-4534-8ff3-9748c2730b9e

**Review by c-ehrlich — APPROVED** [review]

**Comment by benfavre** [comment]
Switching from node 18 to 20 fixes this for me

**Pull request merged**